rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+
+We know that with the current release cadence this is too slow to support TF's rapid development. But with modules, we expect to be able to release much more frequently.
+
+On top of the proposed unit testing plan above, we will need package level integration tests. We propose these to be run every night at head. We propose the following to act as the TF integration tests:
+
+
+
+*   Miscellaneous pip utility tests. Build the nightly pip packages, install them. Then make sure you can import TF, run the command line utilities distributed with TF.
+*   Tutorials/notebook tests. Build the "nightly" pip packages for all above components. Then install this package. Finally extract the Python code from notebooks and run them as graphs
+*   Models tests: Build and install the nightly pip packages, then run curated models under tensorflow/models for a small amount of steps.
+
+The above will simply check the sanity of TF, just will check if TF can run without crashing. However, TF requires much more testing. We propose expansion and adoption of the following regression tests to nightly TF test suites:
+
+*   Convergence tests: Run a curated set of small models until convergence. Measure the time to converge, and steps to converge
+*   Performance tests: Run a curated set of models for a pre-selected number of steps. Measure steps per second, and for image models images per second.
+
+
+## Releases
+
+In the multi package environment, we aim to have the following with the releases:
+
+1.  Smaller packages: We expect to release multiple small packages, instead of the giant package we are building and releasing right now.
+1.  Faster releases: We would like the smaller packages to be able to release much faster. With the ability to pin each dependency to a known good version, we would like to be able to compartmentalize issues, and only hold certain components back when only certain code in TF has issues.
+1.  Independent releases: With compatibility checks, we would like to be able to independently release different packages
+
+Below summarizes a timeline of releases for three packages, A, B and C, where B depends on A and C depends on B.
+
+
+![alt_text](20190305-modular-tensorflow/releases.png "Release plans")
+
+
+To summarize the above timeline:
+
+*   Different packages set their own release cadences
+*   Each package will set version boundaries for each of their dependencies.
+*   Each package is responsible for ensuring that all of their public APIs are working without any changes until the next major release
+*   Packages do not need to modify the minimum version requirements unless they start using newly introduced public API symbols.
+*   TF metapackage releases may choose to hold back individual packages in favor of faster releases. But dependency requirements have to be respected when doing so.
```

Comment:
What Validation will TF Core team do on plugins to ensure interoperability?

Reply:
For a select number of plugins, we will run continuous integration tests. For Filesystems, these will be as simple as reading and writing files. For ops and kernels, we will run convergence and performance testing on select models.

Reply:
Do you know what `select number of plugins` includes?

Reply:
At the moment, "select number of plugins" will be the "features currently officially supported by TF", such as CUDA, cloud file systems, GRPC based networking.
This list may grow or shrink over time.